### DISCUSSION TITLE
Can't set states - only read

### DISCUSSION THREAD

**Discussion by buzzard0627** [tagyoureit/nodejs-poolController#525] [RS485]
Has anyone encounter an issue where they cannot change any states (turn features on or off)?  In looking at my stats, on the RS485 Port, I have a 98.31% failure rate for sending.  Receiving is less than .02% failure.  I've tested and re done the connections multiple times without any change or luck.  I can control everything just fine through the ScreenLogic app.

This is the error I receive when I try to turn something on:
ApiError: Message aborted after 4 attempt(s): 165,49,16,33,134,2,2,1,1,146 
    at RS485Port.writeMessage (/home/adam/nodejs-poolController/controller/comms/Comms.ts:649:27)
    at RS485Port.processWaitPacket (/home/adam/nodejs-poolController/controller/comms/Comms.ts:588:22)
    at RS485Port.processOutboundPackets (/home/adam/nodejs-poolController/controller/comms/Comms.ts:596:19)
    at RS485Port.processPackets (/home/adam/nodejs-poolController/controller/comms/Comms.ts:578:14)
    at Timeout. (/home/adam/nodejs-poolController/controller/comms/Comms.ts:630:52)
    at listOnTimeout (internal/timers.js:557:17)
    at processTimers (internal/timers.js:500:7)

**Comment by rstrouse** [comment]
What are you using as an RS485 adapter.  Is it a TTY adapter by chance?

**Reply by buzzard0627** [reply]
I am using this one: Industrial USB to RS485 Converter Adapter Original FT232RL Fast Communication Embedded Protection Circuits Resettable Fuse ESD Protection TVS Diode etc Automatic Transceiving https://a.co/aqC4xAO

**Reply by rstrouse** [reply]
Are you using this as a network port or is it connected to the ttyUSB0 interface on the pi?

Have you had this working in the past?

**Comment by buzzard0627** [comment]
Direct connect

**Reply by rstrouse** [reply]
Disable the Bluetooth on the pi.  Also, if you have the serial console enabled disable that as well.  I assume this is on a RPi4 or is this another platform?

**Comment by buzzard0627** [comment]
Correct RPI4.

I already had serial disabled.  Same behavior with Bluetooth disabled

**Reply by rstrouse** [reply]
Is the port configuration set to 9600 baud, 8 Data Bits, Parity None, 1 Stop bit?

**Comment by buzzard0627** [comment]
Yep.

**Reply by rstrouse** [reply]
Did this adapter work in the past?

**Reply by buzzard0627** [reply]
Brand new for this purpose.  I’ve already purchased a different model just to test.  Just thought it was strange it could read but not write.

**Reply by rstrouse** [reply]
Shut down njsPC and run npm i in the -/nodejs-poolController directory.  Pay attention to any serial port compile errors.  One sided comms are a potential failure mode for these adapters.

**Reply by buzzard0627** [reply]
No errors outputted during the reinstall, however, it looks like it’s reading less now.  Features, schedules and a few other things are gone.  At this point I’m assuming a bad usb?

**Reply by tagyoureit** [reply]
Any chance you have multiple copies of njsPC running and both accessing the port?  You could try `ps -eo user,pid,cmd -q "$(fuser /dev/ttyS0 2>/dev/null | xargs)"` (change ttyS0 to your adapter) and it will list all applications using the port.  Or `ps -ef | grep node` to see which node processes are running on the system.

**Reply by buzzard0627** [reply]
I don't think I am.  The first command is giving me an error, but the second one is giving me this:
4675  4664  0 07:47 pts/0    00:00:00 sh -c npm run build && node dist/app.js
4706  4695  0 07:48 pts/1    00:00:00 sh -c npm run build && node dist/app.js
4727  4706  0 07:48 pts/1    00:00:03 node dist/app.js
4735  4675  2 07:49 pts/0    00:00:19 node dist/app.js
4785  4652  0 08:04 pts/2    00:00:00 grep --color=auto node

**Reply by tagyoureit** [reply]
What port are you using to access the RS485 adapter?

And multiple node entries will look the same from njsPC, dashPanel and REM. Just make sure there aren't two njsPC instances running.

**Reply by buzzard0627** [reply]
/dev/ttyUSB0

**Reply by tagyoureit** [reply]
> ps -eo user,pid,cmd -q "$(fuser /dev/ttyUSB0 2>/dev/null | xargs)

What error comes back from this command?

**Reply by buzzard0627** [reply]
ps -eo user,pid,cmd -q "$(fuser /dev/ttyUSB0 2>/dev/null | xargs)"
error: List of process IDs must follow -q.

Usage:
 ps [options]

 Try 'ps --help <simple|list|output|threads|misc|all>'
  or 'ps --help <s|l|o|t|m|a>'
 for additional help text.

**Reply by tagyoureit** [reply]
If you get no process id's back, that means no application is accessing the /dev/ttyUSB0 port.

Either your njsPC app is stopped when you were running this or it is pointing to the wrong port.  You config should look like this (with /dev/ttyUSB0).
![2022-07-13_12-37-23](https://user-images.githubusercontent.com/1646840/178817358-4e337657-1ade-4c1a-a4fb-56c0145fc56b.png)

When njsPC is running you should see it (and any other) node processes listed below as item 1.
If you just run the command inside the parens (fuser...) it will show you the pid that is accessing the port as item 2.
When you run the full ps -en... command it should show you that the user (in this case Pi) is accessing my serial port (/dev/ttyS0) through process id (pid) 11450) which I can tie back to the first list of running applications.

![2022-07-13_12-39-12](https://user-images.githubusercontent.com/1646840/178817759-c42ded8c-e711-4560-84e8-95897bd71a8e.png)

**Reply by buzzard0627** [reply]
Weird - not sure.  It is definitely running and based on the UI and running od - x </dev ttyUSB0, I have the right port:

![image](https://user-images.githubusercontent.com/109119227/178820696-0f925963-9dbe-47fb-b48d-f35f4a7ab882.png)

**Comment by buzzard0627** [comment]
Appreciate all the help.  USB device was bad.  New one arrived today and everything started up and was working without any issue.  Thanks for the awesome work!

**Reply by tagyoureit** [reply]
@rstrouse and I keep a bunch on hand... they all seem to go bad sooner or later.  Glad you're up and running again.
